### ISSUE TITLE
Gradle fails with OutOfMemoryError caused by AndroidEntryPointClassTransformer

### ISSUE THREAD

**Issue by mdzyuba** [google/dagger#2567]
Gradle fails with intermittent OutOfMemoryErrors:

```
Caused by: java.lang.OutOfMemoryError: GC overhead limit exceeded
at java.util.Arrays.copyOf(Arrays.java:3181)
at java.util.ArrayList.grow(ArrayList.java:265)
at java.util.ArrayList.ensureExplicitCapacity(ArrayList.java:239)
at java.util.ArrayList.ensureCapacityInternal(ArrayList.java:231)
at java.util.ArrayList.add(ArrayList.java:462)
at javassist.bytecode.ClassFile.addAttribute(ClassFile.java:770)
at javassist.bytecode.ClassFile.read(ClassFile.java:822)
at javassist.bytecode.ClassFile.<init>(ClassFile.java:185)
at javassist.CtClassType.<init>(CtClassType.java:98)
at javassist.ClassPool.makeClass(ClassPool.java:707)
at dagger.hilt.android.plugin.AndroidEntryPointClassTransformer.transformJarContents(AndroidEntryPointClassTransformer.kt:82)
at dagger.hilt.android.plugin.HiltTransformTestClassesTask$WorkerAction.execute(HiltTransformTestClassesTask.kt:84)
```

Build Configuration:
* gradle-6.7.1
* org.gradle.jvmargs=-Xmx8G 
* hiltVersion = '2.31.2-alpha'
* androidPluginVersion=4.2.0-beta02

Based on the heap analysis, it looks like the ClassPool instances are the top consumers of the memory:

![Screen Shot 2021-04-22 at 12 36 41 PM](https://user-images.githubusercontent.com/7852483/115785236-0f1c9f80-a374-11eb-9e43-bac19d7c165b.png)

Note, all the WorkerExecutor Queue Threads in the image above have ClassPool as the main memory consumer. The image just shows the first and last expanded. 

In case it might help with fixing the issue, the ClassPool documentation has a following comment related to the memory consumption and possible optimization options:

 ```
* <p><code>ClassPool</code> objects hold all the <code>CtClass</code>es
 * that have been created so that the consistency among modified classes
 * can be guaranteed.  Thus if a large number of <code>CtClass</code>es
 * are processed, the <code>ClassPool</code> will consume a huge amount
 * of memory.  To avoid this, a <code>ClassPool</code> object
 * should be recreated, for example, every hundred classes processed.
 * Note that <code>getDefault()</code> is a singleton factory.
 * Otherwise, <code>detach()</code> in <code>CtClass</code> should be used
 * to avoid huge memory consumption. 
```

**Comment by danysantiago** [comment]
Thanks for the report. We should totally detach `CtClasses` as they are processed since they don't really need to stay on the pool. 

Also, I am surprise that the `HiltTransformTestClassesTask` is even running if you are using AGP 4.2.0 since that task is only configured on older AGP versions where the transform pipeline didn't affect test sources. Instead, the Plugin should be configuring the build to use a newer more efficient transform pipeline.

**Comment by mdzyuba** [comment]
Thanks, Dany. We have Lint configured to run with AGP 4.2.0 (due to the Lint baseline bug in AGP 4.0.0)  and the rest is on 4.0.0 (due to stability of AGP 4.2.0). The trace listed in the issue report is from AGP 4.0.0 run. 

The upgrades are planned. However, they are somewhat limited by current project dependencies and reliability concerns.

Is there any chance this could be a fix in Hilt 2.31?
